File: ts/test/http-error-format.test.mjs
import test from "node:test";
import assert from "node:assert/strict";

import { createApp, HTTP_ERROR_FORMAT_FLAT_LEGACY } from "../dist/index.js";
import { AppTheoryError } from "../dist/errors.js";

test("default HTTP errors remain nested", async () => {
  const app = createApp({ tier: "p2" });
  app.get("/portable", () => {
    throw new AppTheoryError("VALIDATION_ERROR", "bad input", {
      statusCode: 422,
      details: { field: "config_type" },
    });
  });

  const resp = await app.serve({
    method: "GET",
    path: "/portable",
    headers: { "x-request-id": "req_123" },
    body: "",
  });

  assert.equal(resp.status, 422);
  assert.equal(resp.headers["x-request-id"]?.[0], "req_123");
  assert.deepEqual(JSON.parse(Buffer.from(resp.body).toString("utf8")), {
    error: {
      code: "VALIDATION_ERROR",
      message: "bad input",
      status_code: 422,
      details: { field: "config_type" },
      request_id: "req_123",
    },
  });
});

test("legacy HTTP error format flattens portable and framework errors", async () => {
  const app = createApp({
    tier: "p2",
    httpErrorFormat: HTTP_ERROR_FORMAT_FLAT_LEGACY,
  });
  app.get("/portable", () => {
    throw new AppTheoryError("VALIDATION_ERROR", "bad input", {
      statusCode: 422,
      details: { field: "config_type" },
      traceId: "trace_1",
      requestId: "req_from_error",
    });
  });

  const portable = await app.serve({
    method: "GET",
    path: "/portable",
    headers: { "x-request-id": "req_123" },
    body: "",
  });
  assert.equal(portable.status, 422);
  assert.equal(portable.headers["x-request-id"]?.[0], "req_123");
  assert.deepEqual(JSON.parse(Buffer.from(portable.body).toString("utf8")), {
    code: "VALIDATION_ERROR",
    message: "bad input",
    details: { field: "config_type" },
  });

  const missing = await app.serve({
    method: "GET",
    path: "/missing",
    headers: { "x-request-id": "req_456" },
    body: "",
  });
  assert.equal(missing.status, 404);
  assert.equal(missing.headers["x-request-id"]?.[0], "req_456");
  assert.deepEqual(JSON.parse(Buffer.from(missing.body).toString("utf8")), {
    code: "app.not_found",
    message: "not found",
  });
});

test("legacy HTTP error format applies to HTTP adapter parse failures", async () => {
  const app = createApp({
    tier: "p2",
    httpErrorFormat: HTTP_ERROR_FORMAT_FLAT_LEGACY,
  });

  const resp = await app.serveAPIGatewayV2({
    rawQueryString: "%zz",
    requestContext: { http: { method: "GET", path: "/" } },
  });

  assert.equal(resp.statusCode, 400);
  assert.deepEqual(JSON.parse(resp.body), {
    code: "app.bad_request",
    message: "invalid query string",
  });
});

File: runtime/http_error_format_test.go
package apptheory

import (
	"context"
	"encoding/json"
	"testing"
	"time"

	"github.com/aws/aws-lambda-go/events"
	"github.com/stretchr/testify/require"
)

func TestServe_LegacyHTTPErrorShape_UsesFlatPortableErrorBody(t *testing.T) {
	t.Parallel()

	now := time.Unix(123, 456).UTC()
	app := New(
		WithTier(TierP2),
		WithIDGenerator(fixedIDs{id: "req_generated"}),
		WithLegacyHTTPErrorShape(),
	)
	app.Get("/portable", func(_ *Context) (*Response, error) {
		return nil, NewAppTheoryError("VALIDATION_ERROR", "bad input").
			WithStatusCode(422).
			WithDetails(map[string]any{"field": "config_type"}).
			WithTraceID("trace_1").
			WithTimestamp(now).
			WithRequestID("req_from_error")
	})

	resp := app.Serve(context.Background(), Request{
		Method:  "GET",
		Path:    "/portable",
		Headers: map[string][]string{"x-request-id": {"req_incoming"}},
	})

	require.Equal(t, 422, resp.Status)
	require.Equal(t, []string{"req_incoming"}, resp.Headers["x-request-id"])

	var body map[string]any
	require.NoError(t, json.Unmarshal(resp.Body, &body))
	require.NotContains(t, body, "error")
	require.Equal(t, "VALIDATION_ERROR", body["code"])
	require.Equal(t, "bad input", body["message"])
	require.Equal(t, map[string]any{"field": "config_type"}, body["details"])
	require.NotContains(t, body, "status_code")
	require.NotContains(t, body, "request_id")
	require.NotContains(t, body, "trace_id")
	require.NotContains(t, body, "timestamp")
}

func TestServe_LegacyHTTPErrorShape_AppliesToFrameworkErrors(t *testing.T) {
	t.Parallel()

	app := New(
		WithTier(TierP2),
		WithIDGenerator(fixedIDs{id: "req_generated"}),
		WithLegacyHTTPErrorShape(),
	)

	resp := app.Serve(context.Background(), Request{
		Method:  "GET",
		Path:    "/missing",
		Headers: map[string][]string{"x-request-id": {"req_123"}},
	})

	require.Equal(t, 404, resp.Status)
	require.Equal(t, []string{"req_123"}, resp.Headers["x-request-id"])

	var body map[string]any
	require.NoError(t, json.Unmarshal(resp.Body, &body))
	require.Equal(t, map[string]any{
		"code":    errorCodeNotFound,
		"message": errorMessageNotFound,
	}, body)
}

func TestServeAPIGatewayV2_LegacyHTTPErrorShape_AppliesToAdapterParseErrors(t *testing.T) {
	t.Parallel()

	app := New(WithLegacyHTTPErrorShape())
	out := app.ServeAPIGatewayV2(context.Background(), events.APIGatewayV2HTTPRequest{
		RawQueryString: "%zz",
	})

	require.Equal(t, 400, out.StatusCode)

	var body map[string]any
	require.NoError(t, json.Unmarshal([]byte(out.Body), &body))
	require.Equal(t, errorCodeBadRequest, body["code"])
	require.Equal(t, errorMessageInvalidQueryString, body["message"])
	require.NotContains(t, body, "error")
}
